src/schema-analyzer.ts

```diff
@@ -162,148 +199,166 @@ function fieldComparator(a: SchemaField, b: SchemaField) {
 }
 
 /**
- * Final pass through the result to add missing information:
- *   - compute `probability`, `unique`, `has_duplicates` and
- *     `average_length` fields
- *   - add `Undefined` pseudo-types
- *   - collapse `type` arrays to single string if length 1
- *   - turns fields and types objects into arrays to conform with original
- *     schema parser
- * This mutates the passed in schema.
+ * Returns the type of value as a string. BSON type aware. Replaces `Object`
+ * with `Document` to avoid naming conflicts with javascript Objects.
  */
-function finalizeSchema(schema: any, parent?: any, tag?: 'fields' | 'types'): void {
-  if (schema === undefined) {
-    return;
+function getBSONType(value: any): SchemaBSONType {
+  const bsonType = value?._bsontype
+    ? value._bsontype
+    : Object.prototype.toString.call(value).replace(/^\[object (\w+)\]$/, '$1');
+
+  if (bsonType === 'Object') {
+    // In the resulting schema we rename `Object` to `Document`.
+    return 'Document';
   }
+  return bsonType;
+}
 
-  if (tag === undefined) {
-    // Recursively finalize fields.
-    // debug('recursively calling schema.fields');
-    finalizeSchema(schema.fields, schema, 'fields');
-  }
+function isNullType(type: SchemaAnalysisType): type is SchemaAnalysisNullType {
+  return (<SchemaAnalysisNullType>type).name === 'Null';
```

Comment:
```suggestion
  return (type as SchemaAnalysisNullType).name === 'Null';
```

since that’s how we cast elsewhere (both in general and in the existing code here)? (ditto below)